ophyd/redis_signal.py

```diff
@@ -0,0 +1,306 @@
+
+        if serializer_deserializer is None:
+            self._serializer = json.dumps
+            self._deserializer = json.loads
+        else:
+            self._serializer = serializer_deserializer[0]
+            self._deserializer = serializer_deserializer[1]
+
+        if initial_value is not None:
+            if not self._r.exists(self._key):
+                self.set(initial_value)
+
+        if timestamp is None:
+            timestamp = time.time()
+
+        try:
+            connected = self._r.ping()
+        except redis.ConnectionError:
+            connected = False
+
+        self._metadata = dict(
+            connected=connected,
+            # read_access=True,
+            # write_access=True,
+            timestamp=timestamp,
+            # status=None,
+            # severity=None,
+            # precision=None,
+        )
+
+    @property
+    def timestamp(self):
+        """Timestamp of the readback value"""
+        return self._metadata["timestamp"]
+
+    @property
+    def connected(self):
+        "Is the signal connected to its associated hardware, and ready to use?"
+        return self._metadata["connected"]  # and not self._destroyed
+
+    def set(self, value):
+        """Set value of signal. Sets value of redis key to the serialized dictionary of value and timestamp.
+
+        Returns
+        -------
+        st : Status
+            The status object is set to finished on successful write to redis, or an exception
+            is set if redis.ConnectionError is raised.
+        """
+        st = Status(self)
+        try:
+            server_time = self._r.time()
+            ts = server_time[0] + server_time[1] / 1000000
+            self._r.set(
+                self._key,
+                self._serializer({"value": value, "timestamp": ts}),
+            )
+        except redis.ConnectionError as e:
+            st.set_exception(e)
+        st.set_finished()
+        return st
+
+    def get(self):
+        return self.read()[self.name]["value"]
+
+    def put(self, value):
+        self.set(value)
+
+    def read(self):
+        v = self._r.get(self._key)
+        if v is None:
+            raise NoKey
+
+        self._last_read = self._deserializer(v)
+
+        return {
+            self.name: self._last_read,
+        }
+
+    def describe(self):
+        val = self.read()
+        return {
+            k: {
+                "source": f"redis://{self._r.connection_pool.connection_kwargs['host']}:{self._key}",
+                "dtype": data_type(v["value"]),
+                "shape": data_shape(v["value"]),
+            }
+            for k, v in val.items()
+        }
+
+    def read_configuration(self):
+        return {}
+
+    def describe_configuration(self):
+        return {}
+
+    @property
+    def hints(self):
+        if self.kind == Kind.hinted:
+            return {"fields": [self.name]}
+        else:
+            return {}
+
+    def subscribe(self, *args, **kwargs):
```

Comment:
This should probably only work on `SUB_VALUE` ?

Reply:
This also needs to thread the user's callback through so the users code gets called when the values update.